src/Test/Html/Selector/Internal.elm

```diff
@@ -104,33 +114,60 @@ query :
     -> List (ElmHtml msg)
     -> List (ElmHtml msg)
 query fn fnAll selector list =
-    case selector of
-        All selectors ->
-            fnAll selectors list
+    case list of
+        [] ->
+            list
 
-        Classes classes ->
-            List.concatMap (fn (ElmHtml.Query.ClassList classes)) list
+        elems ->
+            case selector of
+                All selectors ->
+                    fnAll selectors elems
 
-        Class class ->
-            List.concatMap (fn (ElmHtml.Query.ClassList [ class ])) list
+                Classes classes ->
+                    List.concatMap (fn (ElmHtml.Query.ClassList classes)) elems
 
-        Attribute { name, value } ->
-            List.concatMap (fn (ElmHtml.Query.Attribute name value)) list
+                Class class ->
+                    List.concatMap (fn (ElmHtml.Query.ClassList [ class ])) elems
 
-        BoolAttribute { name, value } ->
-            List.concatMap (fn (ElmHtml.Query.BoolAttribute name value)) list
+                Attribute { name, value } ->
+                    List.concatMap (fn (ElmHtml.Query.Attribute name value)) elems
 
-        Style style ->
-            List.concatMap (fn (ElmHtml.Query.Style style)) list
+                BoolAttribute { name, value } ->
+                    List.concatMap (fn (ElmHtml.Query.BoolAttribute name value)) elems
 
-        Tag name ->
-            List.concatMap (fn (ElmHtml.Query.Tag name)) list
+                Style style ->
+                    List.concatMap (fn (ElmHtml.Query.Style style)) elems
 
-        Text text ->
-            List.concatMap (fn (ElmHtml.Query.ContainsText text)) list
+                Tag name ->
+                    List.concatMap (fn (ElmHtml.Query.Tag name)) elems
 
-        Invalid ->
-            []
+                Text text ->
+                    List.concatMap (fn (ElmHtml.Query.ContainsText text)) elems
+
+                Containing selectors ->
+                    let
+                        anyDescendantsMatch elem =
```

Comment:
🎨 I think I would top-level this, since it's already nested.

Reply:
I originally did, but it had to take `selectors` and `fn` and `fnAll` and it seemed cleaner to let it close over those. 😄